### ISSUE TITLE
Arb.stringPattern slows down the test dramatically

### ISSUE THREAD

**Issue by ataronet** [kotest/kotest#1878]
I used to have this in my code 
```
fun Arb.Companion.alphanumeric(): Arb<String> = Arb.stringPattern("[a-zA-Z0-9]+")
```
and used it everywhere. Tests became extremely slow after the upgrade to Kotest 4.3.1 (from 4.2.x).
Let me know if you need more info.

A

**Comment by sksamuel** [comment]
I remember us chatting on slack. I assumed the slowness was just in the library but it only happened in 4.3.x? If so, probably edge case related. Are you using it in binds or just raw ?

**Comment by ataronet** [comment]
@sksamuel I was using it in Bind but then just raw for the investigation. 
Now I have a work around without the pattern.

**Comment by twomoose** [comment]
Hi there. I am also seeing this issue when upgrading from 4.2.6 to 4.3.1
```kotlin
fun Arb.Companion.fooId(arb: Arb<String> = stringPattern("[a-zA-Z0-9:]{1,255}")): Arb<FooId> = arb.map(::FooId)
```
I use `fooId` throughout my test classes and now those tests just hang.

**Comment by myuwono** [comment]
I think i might have found the root cause. `Arb.pattern` is instantiating Generex object on every sample. I managed to replicate this slowdown and confirmed that it is indeed a regression. It seems we need to cache the Generex object. I'll raise a fix for that @sksamuel 

This is the arb https://github.com/kotest/kotest/blob/99138d177bd52feea19740abeec6c5c4a3ae11c1/kotest-property/src/jvmMain/kotlin/io/kotest/property/arbitrary/stringsjvm.kt#L7-L16

```kotlin
fun main() {
    val startCached = TimeSource.Monotonic.markNow()
    arbCachedPattern("[a-zA-Z0-9]+").take(10000).toList()
    println("cached: ${startCached.elapsedNow()}") 
    // cached: 214ms

    val startNotCached = TimeSource.Monotonic.markNow()
    Arb.stringPattern("[a-zA-Z0-9]+").take(10000).toList()
    println("not cached: ${startNotCached.elapsedNow()}")
    // not cached: 2.71s
}

// caching the Generex object
fun arbCachedPattern(pattern: String) = object : Arb<String>() {
    val generex = Generex(pattern)
    override fun edgecases(): List<String> = emptyList()
    override fun sample(rs: RandomSource): Sample<String> {
        generex.setSeed(rs.random.nextLong())
        return sampleOf(generex.random(), StringShrinker)
    }
}
```

**Comment by sksamuel** [comment]
If you make the fix Mitchell I'll release 4.3.2

**Comment by sksamuel** [comment]
4.3.2 is released with this fix.
